### PR TITLE
Avoid leaks due to async exceptions

### LINKED ISSUE #136: packCXString =<< foo pattern may leak
`foo` allocates with malloc. `packCXString` disposes and frees. If async exception comes in between `foo` and `packCXString` some memory is leaked.

### PR DESCRIPTION
**Pull request by edsko**
@phadej I'd appreciate your eyes on this. 

Closes #136.

### CODE REVIEW

**Comment by edsko** [comment]
Critical changes are: 

* definition of `packCXString`
* definition of `wrapForeignPtr` (in `IsForeignPtr` instance for `DeriveIsForeignPtr`)
* definition of `clang_visitChildren`

**Comment by edsko** [comment]
Perhaps we should instead of masking async exceptions _preallocate_ the buffer Haskell side. Especially during the traversal (maybe only there?) that might be easier. I need to think about this some more.

**Comment by phadej** [comment]
> Perhaps we should instead of masking async exceptions preallocate the buffer Haskell side.

Yes, one could have `newtype CXCursor = CXCursor ByteArray`. I do something like that e.g. in `hashable`'s `xxhash` bindings for hash state (https://github.com/haskell-unordered-containers/hashable/blob/57e0d06623d9f2c8a0c5ff13e04b5a4002d699f5/src/Data/Hashable/XXH3.hs#L72), though there state is mutable the concept is the same.

File: hs-bindgen-libclang/src/HsBindgen/Clang/Core.hs [hunk 1]
```diff
@@ -325,7 +327,7 @@ newtype CXCursor_ = CXCursor_ (Ptr ())
 --   AST.
 --
 -- <https://clang.llvm.org/doxygen/structCXCursor.html>
-newtype CXCursor = CXCursor (ForeignPtr CXCursor_)
+newtype CXCursor = CXCursor (ForeignPtr ())
```
Comment (phadej) [inline hs-bindgen-libclang/src/HsBindgen/Clang/Core.hs:330]: is this change necessary?

Comment (edsko) [reply hs-bindgen-libclang/src/HsBindgen/Clang/Core.hs:330]: It brings this in line with all the other definitions, this was just a bug.

Comment (phadej) [reply hs-bindgen-libclang/src/HsBindgen/Clang/Core.hs:330]: It feels that *others* are buggy. We have

```haskell
insance IsForeignPtr CXCursor where
  type UnderlyingPtr CXCursor :: Ptr CXCursor_
  
   unwrapForeignPtr :: forall a. CXCursor -> (Ptr CXCursor_ -> IO a) -> IO a  
```

but we `coerce :: Ptr () -> Ptr CXCursor`, this is ok, there is `castPtr` and the `Ptr` (and `ForeignPtr` arguments are phantom, but it feels wrong.

Comment (edsko) [reply hs-bindgen-libclang/src/HsBindgen/Clang/Core.hs:330]: I don't think so..? 

```haskell
newtype DeriveIsForeignPtr u p = DeriveIsForeignPtr p

instance (Coercible p (ForeignPtr a), Coercible u (Ptr a))
      => IsForeignPtr (DeriveIsForeignPtr u p) where
  type UnderlyingPtr (DeriveIsForeignPtr u p) = u
```

so

```haskell
newtype CXCursor = CXCursor (ForeignPtr CXCursor_)
  deriving IsForeignPtr via DeriveIsForeignPtr CXCursor_ CXCursor
```

results in

```haskell
instance IsForeignPtr CXCursor where
  type UnderlyingPtr CXCursor = CXCursor_
```

not `Ptr CXCursor_`; note that `CXCursor_` is _already_ a `Ptr ()`; `Ptr CXCursor_` would be a pointer to a pointer. 

Am I wrong..?

(I'll experiment with your `ByteArray` suggestion though, which will change this anyway.)

Comment (phadej) [reply hs-bindgen-libclang/src/HsBindgen/Clang/Core.hs:330]: I'd argue that we should have `data CXCursor_` be a tag for the struct. And use `Ptr CXCursor_` to indicate pointer to that struct. `Ptr ()` is writing all bindings with `void *`. Fine, but not precise. `CApiFFI` has less to work with as `void *` pointers can be freely cast.

Comment (edsko) [reply hs-bindgen-libclang/src/HsBindgen/Clang/Core.hs:330]: I'd be okay with that also. I'll look at the `ByteArray` thing and also reconsider this.
